tensorflow/compiler/xla/service/pattern_matcher.h

```diff
@@ -1266,6 +1266,56 @@ class HloInstructionPatternOpcodeImpl {
   bool invert_;
 };
 
+// An HloInstructionPattern implementation that optionally matches a unary
+// operand with a given opcode before matching a given pattern.
+template <typename OperandType, typename OperandImpl>
+class HloInstructionPatternUnaryOptionalOpImpl {
+ public:
+  explicit HloInstructionPatternUnaryOptionalOpImpl(
+      absl::Span<const HloOpcode> opcodes,
+      const HloInstructionPattern<OperandType, OperandImpl>& operand)
+      : opcodes_(opcodes), operand_(operand) {}
+
+  bool Match(::xla::HloInstruction* inst, MatchOption option) const {
+    // In the transparent case, the instruction matches the given operand
+    // pattern.
+    if (operand_.Match(inst, option)) {
+      return true;
+    }
+    // Alternatively, the opcode of the instruction matches one of the entries
+    // of opcodes_.
+    if (!c_linear_search(opcodes_, inst->opcode())) {
+      EXPLAIN << " HloInstruction also doesn't have any of opcodes {"
+              << absl::StrJoin(opcodes_, ", ", opcode_to_string) << "}.";
+      return false;
+    }
+    // Additionally, the operand of the instruction must match the given operand
+    // pattern.
+    if (!operand_.Match(HloOperand(inst, 0), option)) {
+      EXPLAIN
+          << " Matched HloInstruction " << HloOpcodeString(inst->opcode())
+          << ", but its operand differs from the prescribed operand pattern.";
+      return false;
+    }
+    return true;
+  }
+
+  void DescribeTo(std::ostream* os, int64_t indent = 0) const {
+    *os << "which optionally matches a unary operand with any of opcodes {"
+        << absl::StrJoin(opcodes_, ", ", opcode_to_string)
```

Comment:
please inline this lambda (reasons: XLA idiom, fewer variables, putting code closer to where it's used)